File: Problems & Solutions/295. Find Median from Data Stream/12 November 2022/Python/Find Median from Data Stream.py
class MedianFinder:

    def __init__(self):
        self.left = []
        self.right = []

    def addNum(self, num: int) -> None:
        if len(self.left) == len(self.right):
            heappush(self.right, -heappushpop(self.left, -num))
        else:
            heappush(self.left, -heappushpop(self.right, num))
        
    def findMedian(self) -> float:
        if len(self.left) == len(self.right):
            return (-self.left[0] + self.right[0]) / 2
        else:
            return self.right[0]
        


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: Problems & Solutions/295. Find Median from Data Stream/12 November 2022/C++/Find Median from Data Stream.cpp
class MedianFinder {
public:
    int size;
    priority_queue<int> pq1;
    priority_queue<int,vector<int>,greater<int>> pq2;
    MedianFinder() 
    {
           size=0;
    }
    
    void addNum(int num) 
    {
       if(size==0) pq1.push(num);
       else if(size%2==1)
       {
           int x=pq1.top();
           if(num>x) pq2.push(num);
           else 
           {
               pq1.pop();pq1.push(num);
               pq2.push(x);
           }
       }
       else
       {
           int x=pq2.top();
           if(num<x) pq1.push(num);
           else 
           {
               pq2.pop();pq2.push(num);
               pq1.push(x);
           }
       }
     
        size++;
    }
    
    double findMedian() 
    {
       if(size%2==1) return pq1.top();
       else return (double) (pq1.top()+pq2.top())/2;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

